arangod/Cluster/ClusterMethods.cpp

```diff
@@ -753,211 +850,208 @@ bool smartJoinAttributeChanged(LogicalCollection const& collection, VPackSlice c
+    network::Headers headers;
+    auto future =
+        network::sendRequest("shard:" + p.first, fuerte::RestVerb::Get,
+                             "/_db/" + StringUtils::urlEncode(dbname) +
+                                 "/_api/collection/" +
+                                 StringUtils::urlEncode(p.first) + "/revision",
+                             VPackBuffer<uint8_t>(), network::Timeout(300.0), headers);
+    futures.emplace_back(std::move(future));
   }
 
-  // Now listen to the results:
-  int count;
-  int nrok = 0;
-  for (count = (int)shards->size(); count > 0; count--) {
-    auto res = cc->wait(coordTransactionID, 0, "", 0.0);
-    if (res.status == CL_COMM_RECEIVED) {
-      if (res.answer_code == arangodb::rest::ResponseCode::OK) {
-        VPackSlice answer = res.answer->payload();
-
-        if (answer.isObject()) {
-          VPackSlice r = answer.get("revision");
-
-          if (r.isString()) {
-            VPackValueLength len;
-            char const* p = r.getString(len);
-            TRI_voc_rid_t cmp = TRI_StringToRid(p, len, false);
-
-            if (cmp != UINT64_MAX && cmp > rid) {
-              // get the maximum value
-              rid = cmp;
+  auto cb = [rid](std::vector<Try<network::Response>>&& results) -> std::pair<OperationResult, TRI_voc_rid_t> {
+    return handleResponsesFromAllShards<TRI_voc_rid_t>(
+        results, rid, [](Result& result, TRI_voc_rid_t& rid, VPackSlice answer) -> void {
+          if (answer.isObject()) {
+            VPackSlice r = answer.get("revision");
+            if (r.isString()) {
+              VPackValueLength len;
+              char const* p = r.getString(len);
+              TRI_voc_rid_t cmp = TRI_StringToRid(p, len, false);
+
+              if (cmp != UINT64_MAX && cmp > rid) {
+                // get the maximum value
+                rid = cmp;
+              }
             }
+          } else {
+            // didn't get the expected response
+            result.reset(TRI_ERROR_INTERNAL);
           }
-          nrok++;
-        }
-      }
-    }
-  }
-
-  if (nrok != (int)shards->size()) {
-    return TRI_ERROR_INTERNAL;
-  }
-
-  return TRI_ERROR_NO_ERROR;  // the cluster operation was OK, however,
-                              // the DBserver could have reported an error.
+        });
+  };
+  return futures::collectAll(std::move(futures)).thenValue(std::move(cb));
 }
 
-int warmupOnCoordinator(std::string const& dbname, std::string const& cid) {
+futures::Future<OperationResult> warmupOnCoordinator(std::string const& dbname,
+                                                     std::string const& cid) {
   // Set a few variables needed for our work:
   ClusterInfo* ci = ClusterInfo::instance();
   auto cc = ClusterComm::instance();
   if (cc == nullptr) {
     // nullptr happens only during controlled shutdown
-    return TRI_ERROR_SHUTTING_DOWN;
+    return futures::makeFuture(OperationResult(TRI_ERROR_SHUTTING_DOWN));
```

Comment:
I don't think you necessarily need to return a OperationResult. A Result object is also fine no ?

Reply:
I suppose, yes. I chose OperationResult for consistency, but it's somewhat overkill.